### PR TITLE
feat: add async state synchronization, fix tests + fixtures, add docs

### PR DESCRIPTION
**Pull request by bezrazli4n0**

### CODE REVIEW

File: io/Cargo.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
-gstd = { git = "https://github.com/gear-tech/gear.git" }
-scale-info = { version = "2.2.0", default-features = false, features = ["derive"] }
-codec = { package = "parity-scale-codec", version = "3.1.2", default-features = false, features = ["derive"] }
+gstd = { git = "https://github.com/gear-tech/gear.git", features = ["debug"] }
+codec = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
```
Comment (shamilsan) [inline io/Cargo.toml]: No need for features.
```suggestion
codec = { package = "parity-scale-codec", version = "3", default-features = false }
```

File: io/Cargo.toml [hunk 2]
```diff
@@ -4,6 +4,7 @@ version = "0.1.0"
...
+primitive-types = { version = "0.11", default-features = false, features = ["scale-info"] }
+scale-info = { version = "2", default-features = false, features = ["derive"] }
```
Comment (shamilsan) [inline io/Cargo.toml]:
```suggestion
scale-info = { version = "2", default-features = false }
```

File: Cargo.toml [hunk 3]
```diff
@@ -7,8 +7,10 @@ authors = ["Gear Technologies"]
 
 [dependencies]
 gstd = { git = "https://github.com/gear-tech/gear.git" }
-ft-io = { git = "https://github.com/gear-dapps/fungible-token.git", tag = "0.1.0" }
+codec = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
```
Comment (shamilsan) [inline Cargo.toml]:
```suggestion
codec = { package = "parity-scale-codec", version = "3", default-features = false }
```

File: src/lib.rs [hunk 4]
```diff
@@ -37,6 +37,14 @@ impl IcoContract {
     /// * `config`: Consists of `duration`, `start_price`, `tokens_goal`, `price_increase_step` and time_increase_step
     ///
     async fn start_ico(&mut self, config: IcoAction) {
+        let current_transaction_id = *self.transactions.entry(msg::source()).or_insert_with(|| {
```
Comment (shamilsan) [inline src/lib.rs]: I'd keep `msg::source` in a variable instead of calling syscall every time, it will be cheaper.

File: src/messages.rs [hunk 5]
```diff
@@ -1,17 +1,38 @@
...
+/// * `to`: recipient account
+/// * `amount`: amount of tokens
+pub async fn transfer_tokens(
+    transaction_id: u64,
+    token_address: &ActorId,
+    from: &ActorId,
+    to: &ActorId,
+    amount_tokens: u128,
+) -> Result<(), ()> {
+    let reply = msg::send_for_reply_as::<_, FTokenEvent>(
```
Comment (shamilsan) [inline src/messages.rs]: I still prefer avoiding turbo-fish 😄 
```suggestion
    let reply: FTokenEvent = msg::send_for_reply_as(
```

File: Makefile [hunk 6]
```diff
@@ -31,6 +31,20 @@ linter:
...
-	@if [ ! -f "./target/fungible_token-0.1.0.wasm" ]; then wget "https://github.com/gear-dapps/fungible-token/releases/download/0.1.0/fungible_token-0.1.0.wasm" -O "./target/fungible_token-0.1.0.wasm"; fi
+	@if [ ! -f "./target/ft_main.wasm" ]; then\
+	    curl -L\
+	        "https://github.com/gear-dapps/sharded-fungible-token/releases/download/0.1.0/ft_main-0.1.0.opt.wasm"\
+	        -o "./target/ft_main.wasm";\
+	fi
+	@if [ ! -f "./target/ft_logic.opt.wasm" ]; then\
+	    curl -L\
+	        "https://github.com/gear-dapps/sharded-fungible-token/releases/download/0.1.0/ft_logic-0.1.0.opt.wasm"\
+	        -o "./target/ft_logic.opt.wasm";\
```
Comment (shamilsan) [inline Makefile:42]: Why do we keep `.opt.wasm` extension here, when losing it in `ft_main.wasm`?

File: tests/start_ico_tests.rs [hunk 7]
```diff
@@ -72,26 +70,15 @@ fn zero_duration_start_ico() {
...
-        OWNER_ID,
-        InitConfig {
-            name: String::from("MyToken"),
-            symbol: String::from("MTK"),
-            decimals: 18,
-        },
-    );
-
-    assert!(res.log().is_empty());
+    // let ft = Program::ftoken(OWNER_ID, TOKEN_ADDRESS, &sys);
```
Comment (shamilsan) [inline tests/start_ico_tests.rs]: Dead code?

**Review by LouiseMedova — APPROVED** [review]

**Review by shamilsan — APPROVED** [review]

**Pull request merged**